### ISSUE TITLE
Problem with 302 redirect and Auth on initial (but not redirected) link

### ISSUE THREAD

**Issue by karptonite** [guzzle/guzzle#1477]
I am downloading a file with Guzzle. The original url, via https, requires Digest authorization, then returns a 302 redirect to another url on a different domain (http) that does NOT require authorization.

When I try to use Guzzle to download the file following redirects automatically, Guzzle makes the initial connection, gets back a 401, sends again with the correct authorization headers, gets back the 302 redirect--all good. 

Then it tries to send the same exact Authorization headers with the request to the new url, which gets back a 400 bad request, I assume because the redirected URL does not expect authorization.

I have gotten around this by capturing the redirect url via an ```on_stats``` function, setting ```allow_redirects``` to false, and creating a completely separate request for the redirected url. But it seems that there should be a better way to handle this.

**Comment by mtdowling** [comment]
What version are you using? Guzzle should be stripping Auth headers by default when redirecting between domains: https://github.com/guzzle/guzzle/blob/master/src/RedirectMiddleware.php#L189

**Comment by karptonite** [comment]
version 6.2.0.

**Comment by mtdowling** [comment]
Can you provide `debug` output, but omit any sensitive information?

**Comment by karptonite** [comment]
Sure. Here is a gist, with sensitive stuff (and maybe more) redacted. The redacted stuff is the same wherever the [redacted x] labels are the same.
https://gist.github.com/karptonite/0c73e76a14db37814229f7ee4606095d
```'auth' ``` is ```[$user, $password, 'digest']```
Let me know if you need any other details or sample code.  I believe that my explanation in my initial post is correct, but let me know if it isn't clear.

**Comment by karptonite** [comment]
Anyone have any thoughts on this? I managed to hack around it for our use case, but it seems like a legitimate bug that could ensnare other unsuspecting users.

**Comment by kkopachev** [comment]
The problem is that Client when handling digest auth relies on Curl options and sets them inside `$options` array https://github.com/guzzle/guzzle/blob/master/src/Client.php#L348-L352
`RedirectMiddleware` drops Auth header from request, but curl options stay https://github.com/guzzle/guzzle/blob/master/src/RedirectMiddleware.php#L190-L192.

Not sure what would be best fix for that, other then modifying passed-by-reference `$options` array and removing these curl options.

```php
if (defined('CURLOPT_HTTPAUTH')) {
    unset($options['curl'][CURLOPT_HTTPAUTH], $options['curl'][CURLOPT_USERPWD]);
}
```

**Comment by karptonite** [comment]
Anyone get a chance to take a look at this? If nothing else, I feel like this deserves a "bug" flag to draw attention from those who might look for ways to contribute with a pull request.

**Comment by sagikazarmark** [comment]
Agreed, after redirection there should be no authentication if the host is different. If this is not the case, then it is a bug.

**Comment by mcfedr** [comment]
Just repeated this today, 
1st request, Authorization header, Response = 301 in the same origin
2nd request, no Authorization header

As a work around set allow_redirects false and wrote my own loop, not ideal.

**Comment by dennisoderwald** [comment]
+1

**Comment by arjenm** [comment]
+1 indeed

**Comment by sagikazarmark** [comment]
@mcfedr you sure it's the same for you? As far as I can see, this ticket is about *sending* authorization header when the domain is the same, whereas you complain about its lack.

(Or I'm completely missing something....)

**Comment by mcfedr** [comment]
@sagikazarmark I think you are right, read that a bit too quickly. For me it appeared the auth header is not set after a redirect.

**Comment by karptonite** [comment]
I believe this issue still exists, and that it probably shouldn't be automatically closed.

**Comment by karptonite** [comment]
I believe this is still a real issue.

**Comment by kkopachev** [comment]
There is an open PR with a fix: https://github.com/guzzle/guzzle/pull/2845
